File: compiler/feedback/placeholders.go
package feedback

import (
	"wipple/database"
	"wipple/queries"
	"wipple/typecheck"
)

func registerPlaceholders() {
	register(Feedback[typecheck.Type]{
		Id:    "placeholder",
		Rank:  RankPlaceholder,
		Query: queries.Placeholder,
		Render: func(render *Render, node database.Node, ty typecheck.Type) {
			if ty != nil {
				render.WriteString("Found a placeholder of type ")
				render.WriteType(ty)
				render.WriteString(".")
			} else {
				render.WriteString("Found a placeholder.")
			}

			render.WriteBreak()

			if ty != nil {
				render.WriteString("Add a ")
				render.WriteType(ty)
				render.WriteString(" value here before running your program.")
			} else {
				render.WriteString("Add a value here before running your program.")
			}
		},
	})
}

File: compiler/src/feedback/items/placeholders.rs
use crate::{
    feedback::{FeedbackCtx, FeedbackRank, RegisteredFeedback},
    queries,
};

pub fn register(ctx: &mut FeedbackCtx) {
    ctx.register(
        RegisteredFeedback::new(
            "placeholder",
            FeedbackRank::Placeholders,
            queries::placeholder,
            |(node, others, _)| (node.clone(), others.iter().cloned().collect()),
            |w, (_, _, ty)| {
                if let Some(ty) = ty {
                    w.write_string("Found a placeholder of type ");
                    w.write_type(ty.clone());
                    w.write_string(".");
                } else {
                    w.write_string("Found a placeholder.");
                }

                w.write_break();

                if let Some(ty) = ty {
                    w.write_string("Add a ");
                    w.write_type(ty.clone());
                    w.write_string(" value here before running your program.");
                } else {
                    w.write_string("Add a value here before running your program.");
                }
            },
        )
        .show_graph(),
    );
}
